### PR TITLE
Make add missing reference more reliable.

### PR DESCRIPTION
**Pull request by jmarolf**
We now look at the identifier names in an effort to find the IErrorType symbol that triggered the diagnostic. Once we find the error symbol we return the assembly identity assuming that this is the correct error symbol since we can't filter by MissingMetadataTypeSymbol.

fixes internal bug 1166387

### CODE REVIEW

**Comment by jmarolf** [comment]
@srivatsn @tmeschter @shyamnamboodiripad @heejaechang @mavasani @JohnHamby Could you take a look please?

File: src/Features/CSharp/CodeFixes/AddMissingReference/AddMissingReferenceCodeFixProvider.cs [hunk 1]
```diff
@@ -3,15 +3,16 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Composition;
-using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.AddMissingReference;
 using Microsoft.CodeAnalysis.CodeFixes;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
```
Comment (tmeschter) [inline src/Features/CSharp/CodeFixes/AddMissingReference/AddMissingReferenceCodeFixProvider.cs]: Fix the order of usings.

**Comment by jmarolf** [comment]
ping @srivatsn @tmeschter @shyamnamboodiripad @heejaechang @mavasani @JohnHamby

File: src/Features/Core/AddMissingReference/AbstractAddMissingReferenceCodeFixProvider.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            return node.DescendantNodesAndSelf().OfType<TIdentifierNameSyntax>();
+        }
+
+        private static IEnumerable<ITypeSymbol> GetTypesForNodes(SemanticModel model, IEnumerable<SyntaxNode> nodes, CancellationToken cancellationToken)
+        {
+            foreach (var node in nodes)
+            {
+                yield return model.GetTypeInfo(node, cancellationToken).Type;
+                var symbol = model.GetSymbolInfo(node, cancellationToken).GetAnySymbol();
+                if (symbol != null)
```
Comment (srivatsn) [inline src/Features/Core/AddMissingReference/AbstractAddMissingReferenceCodeFixProvider.cs]: Can you please add comments explaining what's happening here? Why are we looking at the symbol's parameters? Do we need to look at return types?

File: src/Features/Core/AddMissingReference/AbstractAddMissingReferenceCodeFixProvider.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                    {
+                        foreach (var param in (symbol as IPropertySymbol).Parameters)
+                        {
+                            yield return param.Type;
+                        }
+                    }
+                    yield return symbol?.GetContainingTypeOrThis();
+                }
+                var originalSymbol = symbol?.OriginalDefinition;
+                if (originalSymbol != null)
```
Comment (srivatsn) [inline src/Features/Core/AddMissingReference/AbstractAddMissingReferenceCodeFixProvider.cs]: Refactor code below into a method and just pass in originalSymbol?

**Comment by jmarolf** [comment]
@dpoeschl @Pilchie @jasonmalinowski @balajikris @rchande @brettfo Would anyone from the IDE team be willing to have a look?

**Comment by srivatsn** [comment]
:+1:

File: src/Features/Core/AddMissingReference/AbstractAddMissingReferenceCodeFixProvider.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+        /// <summary>
+        /// Look for additional symbols related to the symbol that we have.
+        /// All of these are candidates for the IErrorTypeSymbol that is causing the missing assembly reference.
+        /// </summary>
+        private static IEnumerable<ITypeSymbol> GetTypesFromSymbol(ISymbol symbol)
+        {
+            if (symbol != null)
+            {
+                if (symbol is IMethodSymbol)
```
Comment (jasonmalinowski) [inline src/Features/Core/AddMissingReference/AbstractAddMissingReferenceCodeFixProvider.cs]: There's a ISymbolExtensions.GetParameters that checks if it's a method or property symbol and returns the appropriate stuff.

File: src/Features/Core/AddMissingReference/AbstractAddMissingReferenceCodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            return symbols;
+        }
+
+        /// <summary>
+        /// Look for additional symbols related to the symbol that we have.
+        /// All of these are candidates for the IErrorTypeSymbol that is causing the missing assembly reference.
+        /// </summary>
+        private static IEnumerable<ITypeSymbol> GetTypesFromSymbol(ISymbol symbol)
+        {
+            if (symbol != null)
```
Comment (jasonmalinowski) [inline src/Features/Core/AddMissingReference/AbstractAddMissingReferenceCodeFixProvider.cs]: Isn't really necessary -- your is checks will just fail already.

**Comment by jasonmalinowski** [comment]
Late to the party, but :+1:.

**Pull request merged**